kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -456,12 +456,16 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
         // Create node upfront -- for common cases it just initializes JobNode.job field,
         // for user-defined handlers it allocates a JobNode object that we might not need, but this is Ok.
         val node: JobNode = makeNode(handler, onCancelling)
+        var nodeAdded: JobNode? = null // prevents handler from double invocation
         loopOnState { state ->
             when (state) {
                 is Empty -> { // EMPTY_X state -- no completion handlers
                     if (state.isActive) {
                         // try move to SINGLE state
-                        if (_state.compareAndSet(state, node)) return node
+                        if (_state.compareAndSet(state, node)) {
+                            nodeAdded = node
+                            return node
```

Comment:
No need to update `nodeAdded` here as we immediately return from this function. These lines can be left intact